src/service/VotingUtils.ts

```diff
@@ -0,0 +1,82 @@
+
+export class VotingUtils {
+    public static insert(result: Uint8Array, value: Uint8Array, index: number): number {
+        result.set(value, index);
+        return index + value.length;
+    }
+    public static createVotingFile(secret: string, votingKeyStartEpoch: number, votingKeyEndEpoch: number): Uint8Array {
+        const items = votingKeyEndEpoch - votingKeyStartEpoch + 1;
+        const headerSize = 64 + 16;
+        const itemSize = 32 + 64;
+        const totalSize = headerSize + items * itemSize;
+        const keyPair = KeyPair.createKeyPairFromPrivateKeyString(secret);
+        const result = new Uint8Array(totalSize);
+        //start-epoch (8b),
+        let index = 0;
+        index = this.insert(result, Convert.numberToUint8Array(votingKeyStartEpoch, 8), index);
+
+        //end-epoch (8b),
+        index = this.insert(result, Convert.numberToUint8Array(votingKeyEndEpoch, 8), index);
+
+        // could it have other values????
+        //last key identifier (8b) - for fresh file this is 0xFFFF'FFFF'FFFF'FFFF (a.k.a. Invalid_Id)
+        index = this.insert(result, Convert.hexToUint8('FFFFFFFFFFFFFFFF'), index);
+
+        //last wipe key identifier (8b) - again, for fresh file this is 0xFFFF'FFFF'FFFF'FFFF (Invalid_Id)
+        index = this.insert(result, Convert.hexToUint8('FFFFFFFFFFFFFFFF'), index);
+
+        // root public key (32b) - this is root public key that is getting announced via vote link tx
+        index = this.insert(result, keyPair.publicKey, index);
+        // start-epoch (8b), \ those two are exactly same one, as top level, reason is this was earlier a tree,
+        index = this.insert(result, Convert.numberToUint8Array(votingKeyStartEpoch, 8), index);
+
+        //end-epoch (8b), / and each level holds this separately, so we left it as is
+        index = this.insert(result, Convert.numberToUint8Array(votingKeyEndEpoch, 8), index);
+        /// what follows are bound keys, there are (end - start + 1) of them.
+
+        // each key is:
+        for (let i = 0; i < items; i++) {
+            // random PRIVATE key (32b)
+            const randomPrivateKey = Crypto.randomBytes(32);
+            const randomKeyPar = KeyPair.createKeyPairFromPrivateKeyString(Convert.uint8ToHex(randomPrivateKey));
+            index = this.insert(result, randomPrivateKey, index);
+            // signature (64b)
+            // now the signature is usual signature done using ROOT private key on a following data:
+            //   (public key (32b), identifier (8b))
+            //
+            //   identifier is simply epoch, but, most importantly keys are written in REVERSE order.
+            //
+            //   i.e. say your start-epoch = 2, end-epoch = 42
+            const identifier = Convert.numberToUint8Array(votingKeyEndEpoch - i, 8);
+            const signature = KeyPair.sign(randomKeyPar, Uint8Array.from([...randomKeyPar.publicKey, ...identifier]));
+            index = this.insert(result, signature, index);
```

Comment:
this is wrong, signature is done using ROOT priv key

Reply:
Yeap, found it in my second commit. Thanks!